### PR TITLE
[Merged by Bors] - feat: some improvements to `apply_fun`

### PR DESCRIPTION
**Pull request by alexjbest**
- Re-enabling applying an order equiv to the target
- using `withMainContext` so that locals are treated properly
- enabling some commented out tests

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/ApplyFun.lean [hunk 1]
```diff
@@ -169,8 +172,8 @@ placeholders. Named placeholders (like `?a` or `?_`) will produce new goals.
  -/
 syntax (name := applyFun) "apply_fun " term (location)? (" using " term)? : tactic
 
-elab_rules : tactic | `(tactic| apply_fun $f $[$loc]? $[using $P]?) => do
-  let P ← P.mapM (elabTerm · none)
+elab_rules : tactic | `(tactic| apply_fun $f $[$loc]? $[using $P]?) => withMainContext do
```
Comment (alexjbest) [inline Mathlib/Tactic/ApplyFun.lean:175]:
```suggestion
elab_rules : tactic | `(tactic| apply_fun $f $[$loc]? $[using $P]?) => do
```

File: Mathlib/Tactic/ApplyFun.lean [hunk 2]
```diff
@@ -169,7 +172,7 @@ placeholders. Named placeholders (like `?a` or `?_`) will produce new goals.
  -/
 syntax (name := applyFun) "apply_fun " term (location)? (" using " term)? : tactic
 
-elab_rules : tactic | `(tactic| apply_fun $f $[$loc]? $[using $P]?) => do
+elab_rules : tactic | `(tactic| apply_fun $f $[$loc]? $[using $P]?) => withMainContext do
   let P ← P.mapM (elabTerm · none)
```
[the remark this answers is not in the document]
Comment (alexjbest) [reply Mathlib/Tactic/ApplyFun.lean]: I've merged master and fixed alias

**Review by kmill — COMMENTED** [review]
Thanks. Once mathlib builds, let's merge!

bors d+

File: test/apply_fun.lean [hunk 3]
```diff
@@ -122,47 +120,39 @@ example (a b : List α) (P : a = b) : True := by
...
---   -- (previously `apply_fun` was producing `f (A.mul B) = f 0`)
---   guard_hyp' h : f (A * B) = f 0,
---   exact h,
--- end
+example {n : Type} [Fintype n] {X : Type} [Semiring X]
+  (f : Matrix n n X → Matrix n n X) (A B : Matrix n n X) (h : A * B = 0) : f (A * B) = f 0 := by
+  apply_fun f at h
+  -- check that our β-reduction didn't mess things up:
+  -- (previously `apply_fun` was producing `f (A.mul B) = f 0`)
+  guard_hyp h : f (A * B) = f 0
```
[the remark this answers is not in the document]
Comment (alexjbest) [reply test/apply_fun.lean]: I made it a syntactic guard hyp

**Pull request closed without merging**